### PR TITLE
Add berth switch reservations

### PR DESCRIPTION
**Pull request by Frodotus**
Adds berth switch reservations
Refs VEN-22

### CODE REVIEW

File: reservations/admin.py [hunk 1]
```diff
@@ -19,6 +19,23 @@
...
+class ReservationTypeFilter(admin.SimpleListFilter):
+    title = _("Reservation type")
+
+    parameter_name = "berth_switch"
+
+    def lookups(self, request, model_admin):
+        return (("0", _("Reservation")), ("1", _("Switch")))
+
+    def queryset(self, request, queryset):
+        kwargs = {"%s" % self.parameter_name: None}
```
Comment (igordavydsson) [inline reservations/admin.py]: let's use the preferred Python 3 way of formatting strings
```python
kwargs = {"{}".format(self.parameter_name): None}
```

File: reservations/models.py [hunk 2]
```diff
@@ -17,6 +17,14 @@ class Meta:
         unique_together = ("reservation", "priority")
 
 
+class BerthSwitch(models.Model):
+    harbor = models.ForeignKey(Harbor, on_delete=models.CASCADE)
+    pier = models.CharField(verbose_name=_("pier"), max_length=40, blank=True)
+    berth_number = models.CharField(
+        verbose_name=_("berth number"), max_length=20, blank=False
```
Comment (igordavydsson) [inline reservations/models.py]: let's do this non-blankable.

File: reservations/models.py [hunk 3]
```diff
@@ -33,6 +41,14 @@ class Reservation(models.Model):
         Harbor, through=HarborChoice, verbose_name=_("chosen harbors"), blank=True
     )
 
+    berth_switch = models.ForeignKey(
+        BerthSwitch,
+        verbose_name=_("harbor switch"),
```
Comment (igordavydsson) [inline reservations/models.py]: `verbose_name=_("berth switch"),` for consistency

File: reservations/schema.py [hunk 4]
```diff
@@ -14,11 +14,22 @@ class Meta:
...
+        model = BerthSwitch
+
+
 class HarborChoiceInput(graphene.InputObjectType):
     harbor_id = graphene.String()
     priority = graphene.Int()
 
 
+class BerthSwitchInput(graphene.InputObjectType):
+    harbor_id = graphene.String()
```
Comment (igordavydsson) [inline reservations/schema.py]: maybe `graphene.Int()`?

File: reservations/schema.py [hunk 5]
```diff
@@ -58,18 +69,34 @@ class ReservationInput(graphene.InputObjectType):
 class CreateReservation(graphene.Mutation):
     class Arguments:
         reservation = ReservationInput()
+        switch = BerthSwitchInput()
```
Comment (igordavydsson) [inline reservations/schema.py]: `berth_switch` for consistency again

File: reservations/admin.py [hunk 6]
```diff
@@ -94,9 +111,13 @@ class ReservationAdmin(admin.ModelAdmin):
             },
         ),
     ]
-    list_display = ("created_at", "first_name", "last_name")
+    list_display = ("created_at", "first_name", "last_name", "switch")
+    list_filter = (ReservationTypeFilter,)
     actions = ["export_reservations", "resend_reservation_confirmation"]
 
+    def switch(self, obj):
```
Comment (igordavydsson) [inline reservations/admin.py]: maybe `switch_application` would be more descriptive? also, can we make this dynamic field translatable?

File: reservations/schema.py [hunk 7]
```diff
@@ -58,18 +69,34 @@ class ReservationInput(graphene.InputObjectType):
...
         reservation_data = kwargs.pop("reservation")
+        switch_data = kwargs.pop("switch")
         try:
             obj_id = reservation_data.pop("boat_type")
         except KeyError:
             obj_id = None
         if obj_id:
             reservation_data["boat_type"] = BoatType.objects.get(id=obj_id)
+
+        try:
```
Comment (igordavydsson) [inline reservations/schema.py]: getting a bit hard to read with all these try blocks. if you have time/energy, you could refactor this maybe, or I will refactor it later at some point.

File: reservations/admin.py (hunk 6 above)
Comment (Frodotus) [reply reservations/admin.py]: Changed it to reservation type, is that ok?

File: reservations/admin.py [hunk 8]
```diff
@@ -94,9 +111,13 @@ class ReservationAdmin(admin.ModelAdmin):
             },
         ),
     ]
-    list_display = ("created_at", "first_name", "last_name")
+    list_display = ("created_at", "first_name", "last_name", "reservation_type")
+    list_filter = (ReservationTypeFilter,)
     actions = ["export_reservations", "resend_reservation_confirmation"]
 
+    def reservation_type(self, obj):
+        return _('Reservation') if obj.berth_switch is None else _('Switch application')
```
Comment (igordavydsson) [inline reservations/admin.py]: Very nice!

File: reservations/schema.py [hunk 9]
```diff
@@ -58,6 +69,7 @@ class ReservationInput(graphene.InputObjectType):
 class CreateReservation(graphene.Mutation):
     class Arguments:
         reservation = ReservationInput()
+        berth_switch = BerthSwitchInput(required=False)
```
Comment (igordavydsson) [inline reservations/schema.py]: `required=False` is deffault behavior, so you could omit that. might be a good idea to add `required=True` to `ReservationInput()` though

**Pull request merged**